kirkwood: fix mtdparts in Pogo E02

bobafetthotmail opened the pull request:
the Pogoplug E02 patch was missing the
correct mtdparts for this device so it
was using the upstream mtd partitions
and could not boot OpenWrt images.

Add the right mtdparts to boot OpenWrt.

Signed-off-by: Alberto Bursi <bobafetthotmail@gmail.com>

Tested on a Pogoplug E02.

CHKDSK88 commented:
I can't find those changes in history. Is 4.19, 4,14 and stable releases affected?

adrianschmutzler commented:
I just checked quickly and found that this is quite old, here the "wrong" partitioning was added for kernel 3.18 in 2014:
https://github.com/openwrt/openwrt/commit/9a7402620684feed7b8f3501a24d525e1df7b7c0

So, I hesitate to believe that this has been wrong for so many years and nobody realized it, particularly if the consequences are as harsh as reported here?

bobafetthotmail commented:
> I can't find those changes in history. Is 4.19, 4,14 and stable releases affected?

Yes. They take upstream mtdparts and they don't work for obvious reasons.

bobafetthotmail commented:
> So, I hesitate to believe that this has been wrong for so many years and nobody realized it, particularly if the consequences are as harsh as reported here?

Kirkwood target has been a mess until after the LEDE split, and it has no "maintainer" even now. There was dead code for making an image for NSA310S, the code for making firmware images was weird and outdated, and so on. During LEDE period this weirdness was fixed up by **nbd** and then a bit more by me and **pwussi** but some devices were left out as we didn't have hardware to test this.
Pogo E02 is one of the first gen kirkwood devices.

That said, facts are facts, the upstream dts file has these mtdparts  https://github.com/torvalds/linux/blob/6f0d349d922ba44e4348a17a78ea51b7135965b1/arch/arm/boot/dts/kirkwood-pogo_e02.dts

```
	partition@0 {
		label = "u-boot";
		reg = <0x0000000 0x100000>;
		read-only;
	};

	partition@100000 {
		label = "uImage";
		reg = <0x0100000 0x400000>;
	};

	partition@500000 {
		label = "pogoplug";
		reg = <0x0500000 0x2000000>;
	};

	partition@2500000 {
		label = "root";
		reg = <0x02500000 0x5b00000>;
	};
```

Which are more or less the stock firmware partitions and will not work with OpenWrt.

These devices are supported by other two projects (ALARM aka Arch linux on ARM and Bodhi's Debian for Kirkwood) too, so most people just went with that instead.

Another workaround is to install the uboot from these projects (as both provide a custom uboot) and then change the configuration so that it passes the correct mtdparts on the kernel commandline. I did help a guy on the forum (that was already running Arch on his Pogo) to do just this. https://forum.openwrt.org/t/upgrade-pogoplug-mobile-or-series-4-from-openwrt-to-lede/9881/60?u=bobafetthotmail

bobafetthotmail commented:
Btw, you can also see how the above mtdparts don't match what was written as "openwrt mtdparts" in the old archived wiki as well (this is the read-only version, last modified in 2015) https://oldwiki.archive.openwrt.org/toh/cloudengines/pogoplug

(that was more or less copy-pasted in the current wiki https://openwrt.org/toh/cloudengines/pogoplug )

mans0n commented:
The partition table is the same as [uboot-kirkwood](https://github.com/openwrt/openwrt/blob/master/package/boot/uboot-kirkwood/patches/140-pogoplug_e02.patch), so I guess this is right. What I don't understand is why it was wrong all the time.

adrianschmutzler commented:
So, this patch added the "correct" config in uboot-kirkwood:
https://github.com/openwrt/openwrt/commit/d3fc4fbd74dcd4dbeaa29d222337eeb6c187a4fd

And this patch added the DTS which was wrong from the very beginning:
https://github.com/openwrt/openwrt/commit/2b0fa00da8fc26e34f08680670f4637766bbd3e8

I'd welcome to have this information in the commit message as well (1. DTS was wrong from the beginning, 2. uboot has "correct" partitions, maybe with commit references)

@mans0n @CHKDSK88 Feel free to review and provide a Reviewed-by .

bobafetthotmail commented:
ok, I'll wait a bit if anyone of you want to add a "revewed by" and then I'll update the commit with the required information.

CHKDSK88 commented:
@adrianschmutzler 
Are we work now only at 5.4? Or @bobafetthotmail should prepare 4.1x patches?

adrianschmutzler commented:
Since we are on 5.4 (without testing) already, I'd accept both.

CHKDSK88 commented:
LGTM.

The wiki should be corrected and IMO it should be backported to 18.06 and 19.07 because current images must be very hard to use.

Reviewed-by: Pawel Dembicki <paweldembicki@gmail.com>

adrianschmutzler commented:
If backporting is desired, I'd actually prefer to have at least 4.14 patches in this PR already.

mans0n commented:
Sorry for my late reply.
I've never used uboot-kirkwood, so at first I thought I'm not really qualified to provide a Reviewed-by here.
But now I come to think of it, the upstream partiton table does not make sense in OpenWrt anyway, so I guess we don't have reason not to accept the patch.

`Reviewed-by: Sungbo Eo <mans0n@gorani.run>`

And just one comment below:

mans0n commented on target/linux/kirkwood/patches-5.4/110-pogo_e02.patch:
```diff
@@ -30,3 +30,39 @@
+-		label = "u-boot";
+-		reg = <0x0000000 0x100000>;
++		label = "uboot";
++		reg = <0x0000000 0xe0000>;
```
If we are going to use 0xe0000 in uboot_env, why don't we use just 0x0 here?

bobafetthotmail replied:
what do you mean? where you want to use 0x0?

mans0n replied:
I mean, in replace of 0x0000000.

bobafetthotmail replied:
that's cosmetic. It is correct either way.

mans0n replied:
Yes, it's cosmetic, but I wanted it to be consistent. I won't mind it if @adrianschmutzler is also okay with it.

adrianschmutzler replied:
@bobafetthotmail 

Either no leading zeros for all partitions (i.e. changing to 0x0 here) or the same number of digits (=7) for all. Currently, it's mixed, and that's misleading on a quick look.

Depending on your decision, I can also adjust this during merge.

adrianschmutzler commented:
@bobafetthotmail So you got two reviewed-by and one question. Please update and add 4.19 and 4.14.

bobafetthotmail commented:
please wait, the patch for kernel 4.14 isn't right

bobafetthotmail commented:
ok, updated correctly.

Also all three kernels have been compiled and run tested on the device.

adrianschmutzler commented on target/linux/kirkwood/patches-4.14/110-pogo_e02.patch:
```diff
@@ -30,3 +30,43 @@
+-	partition@100000 {
+-		label = "uImage";
+-		reg = <0x0100000 0x400000>;
++	partition@0xe0000 {
```
Ah, and the syntax is partition@hexoffset, so not `partition@0xe0000` but `partition@e0000`
Please change that for all partitions and kernels.

bobafetthotmail committed:
kirkwood: fix pogoplug E02 mtdparts

the Pogoplug E02 patch was missing the
correct mtdparts for this device so it
was using the upstream mtd partitions
and could not boot OpenWrt images.

The patch that added the correct config
in uboot-kirkwood is:
uboot-kirkwood: re-add Pogoplug E02 support
d3fc4fbd74dcd4dbeaa29d222337eeb6c187a4fd

The patch that added the current (wrong)
DTS is:
kirkwood: add Pogoplug E02 Kernel support
2b0fa00da8fc26e34f08680670f4637766bbd3e8

Correct the mtdparts for kernel
4.14, 4.19 and 5.4

Reviewed-by: Pawel Dembicki <paweldembicki@gmail.com>
Reviewed-by: Sungbo Eo <mans0n@gorani.run>

Signed-off-by: Alberto Bursi <bobafetthotmail@gmail.com>

bobafetthotmail commented:
aaand updated again with the latest feedback

adrianschmutzler commented:
I'm build-testing right now and will merge it later if no problems occur.
Thanks.
